doc: split up SubmittingPatches.rst

smithfarm opened the pull request:
(https://github.com/ceph/ceph/pull/22876 was too cluttered, so I replaced it with this fresh PR)

Split the existing SubmittingPatches.rst into three files:

* `SubmittingPatches.rst` (general guidelines, for GitHub/master branch)
* `SubmittingPatches-kernel.rst` (for kernel patches)
* `SubmittingPatches-backports.rst` (for backports)

Fixes: http://tracker.ceph.com/issues/20953

---

TODO:

- [x] fix internal hyperlinks
- [ ] after merge, link to SubmittingPatches-backports.rst from https://tracker.ceph.com/projects/ceph-releases/wiki/HOWTO_backport_commits

smithfarm commented:
Rendered versions of the files can be seen at:

https://github.com/smithfarm/ceph/blob/wip-sp-backporting-2/SubmittingPatches.rst
https://github.com/smithfarm/ceph/blob/wip-sp-backporting-2/SubmittingPatches-kernel.rst
https://github.com/smithfarm/ceph/blob/wip-sp-backporting-2/SubmittingPatches-backports.rst

liewegas commented on SubmittingPatches.rst:
```diff
@@ -62,389 +67,108 @@ then you just add a line saying ::
-generates release notes and it is tedious to clean up non-conformant PR titles
-at release time.  This document places no limit on the length of PR titles, but
-be aware that they are subject to editing as part of the release process.
+ you believe your changes should be backported to stable branches after the PR
```
missing "If" ?

smithfarm commented:
@liewegas Thanks for the review. The typo should be fixed now, and I also pushed a second commit which adds a "Test your changes" section to `SubmittingPatches.rst` if you care to take a look.

idryomov commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+Once the PR targeting master is open, add the PR number assigned by GitHub to
+the tracker issue. For example, if the PR number is 99999::
+
+    Pull request ID: 999999
```
Extra 9 here.

idryomov commented on SubmittingPatches.rst:
```diff
@@ -62,468 +66,207 @@ then you just add a line saying ::
+For example::
 
-     osd: make the ClassHandler::mutex private
+     Fixes: 9dbe7a003989f8bb45fe14aaa587e9d60a392727
```
I don't we have ever practised this, at least not in the main repo.  Not saying that this is a bad idea, just that it doesn't codify existing practice and should be discussed on the mailing list first for it to get any traction.  Because everyone is doing their own thing: I noticed @batrick has been using Introduced-by, I usually put an abbreviated SHA1 along with the title somewhere in the commit message, etc.

liewegas replied:
We recently started making a point of mentioning the commit that introduced the problem to help identify follow-on fixes to backport, but I don't think we standardized on the Prefix: that precedes it.  Fixes: seems appropriate though?  In any case, this is a good time to go ahead codify this!

smithfarm replied:
@idryomov This format was suggested by Sage quite some months (or even a year or more) ago.

gregsfortytwo replied:
I'm down with this format.

batrick replied:
+1

smithfarm commented:
@idryomov Thanks for the review. I think I have addressed all the comments.

tchaikov commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+Before you open that PR, please read (at the very least) the following general
+rules::
+
+Backports should state (either in the backport PR or in the tracker issue):
```
what is "the tracker issue" ? the original one or the backport one?

tchaikov commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+
+1. what bug is fixed
+2. why this fix is the minimal way to do it
+3. why does this need to be fixed in <release>
```
i agree that we should always take these into consideration when backporting a fix. but i am not sure we should/could enforce these.

smithfarm replied:
These are guidelines that @athanatos suggested long ago. Note that I wrote "backports should state", not "backports *must* state". I am not suggesting that we start enforcing this in the sense of "rejecting backports that don't include explicit statements satisfying these guidelines", but I think that these guidelines are a good starting point when considering whether, and how, to backport something.

tchaikov replied:
but, i think unlike guidelines, "rules" are supposed to be followed.

tchaikov commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+* all bugfixes and features should land in master first
+* commits to stable branches should be cherry-picked from master
+* cherry-picks must be done using ``git cherry-pick -x``
+* if a commit cannot be cherry-picked from master, the commit message must explain why that was not possible
```
s/was/is/

smithfarm replied:
I intended for this sentence to be in past tense, since the commit message is written *after* the cherry-picking is done. But you're right - it's not consistent. I'll do `s/cannot/could not/`

tchaikov commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+* cherry-picks must be done using ``git cherry-pick -x``
+* if a commit cannot be cherry-picked from master, the commit message must explain why that was not possible
+* the commit message generated by ``git cherry-pick -x`` must not be modified, except to add a "Conflicts" section
+* the "Conflicts" section must mention all files where changes had to made manually (not just conflicts flagged by git)
```
s/had to/have to be/

tchaikov commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+    Status: Pending Backport
+
+Once all of the above has been completed, issues can be created in the Backport
+tracker for tracking the backporting work. In our example, above, two issues
```
```suggestion
tracker for tracking the backporting work. In the above example, two issues
```

smithfarm replied:
@tchaikov "In the example above," <- would that be OK for you?

(Since "above" is not an adjective, strictly speaking it should not go before the noun, though I'm sure you can find examples of such usage on the net.)

tchaikov commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+Python 3 should already be present on any recent Linux installation. The second
+dependency, python-redmine, can be obtained from PyPi::
+
+    https://pypi.org/project/python-redmine/
```
how about just noting down the command to install it
```sh
pip install --user python-redmine
```

tchaikov commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+
+* all bugfixes and features should land in master first
+* commits to stable branches should be cherry-picked from master
+* cherry-picks must be done using ``git cherry-pick -x``
```
this line duplicated with the one on line 22
```rst
* cherry-picks must be done using ``git cherry-pick -x``
```

is this intended?

the same applies to the following lines.

tchaikov commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+backport branch yourself and use the script only for creating the PR and
+updating the Backport tracker issue.
+
+If you prefer to leave everything to the script, just provide the Backport
```
```suggestion
If you prefer leaving everything to the script, just provide the Backport
```

smithfarm replied:
@tchaikov "prefer" can take a gerund or an infinitive - see https://www.wordsmyth.net/blog/2013/12/prefer/

To a native speaker, there is a subtle difference.

tchaikov commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+
+    wip-55555-nautilus
+
+If you prefer to create the backport branch yourself, just do that. Be sure to
```
```suggestion
If you prefer creating the backport branch yourself, just do that. Be sure to
```

wjwithagen commented on SubmittingPatches.rst:
```diff
@@ -2,32 +2,36 @@
-percolate to their final resting place in the kernel through several
-layers of maintainers, we've introduced a "sign-off" procedure on
-patches that are being emailed around.
+Sign your work
```
Perhaps you'd like to suggest/indicate what needs to go in the headers of new files if one creates them... I usually just borrow some content of a file next to it in the directory. 
But that is not always fitting.
I'm thinking that there should be something there in the lines that the file falls under the same copyright as the project.
Also what to do with the original author. Because I'm borrowing, most of the attribution goes to Dreamhost or RedHat.

tchaikov commented on SubmittingPatches.rst:
```diff
@@ -62,468 +66,207 @@ then you just add a line saying ::
-connect this change to addressed issue(s). So a line saying ::
+When you open a GitHub PR with multiple patches, the patches should all be
+related in some way and the PR title should state briefly what the patches do
+and why they are needed. In other words, the patch series and its description
```
i don't think the title should state the "why" part. the body of a commit message should.

tchaikov commented on SubmittingPatches.rst:
```diff
@@ -62,468 +66,207 @@ then you just add a line saying ::
+-----------------
 
-So a typical commit message for revising the document could look like::
+Before you open your PR, it's a good idea to run tests on your patchset. Doing
```
```suggestion
Before opening your PR, it's a good idea to run tests on your patchset. Doing
```

batrick commented on SubmittingPatches.rst:
```diff
@@ -62,468 +66,207 @@ then you just add a line saying ::
-Tag the commit
-^^^^^^^^^^^^^^
+There's nothing inherently wrong with a long commit message body. But it might
+be a sign that the patch could benefit from being split up into smaller patches.
```
I think we want to encourage long explanatory commit messages and this paragraph seems to discourage it. A long commit message may include error output (from teuthology) or other useful information to give background for the patch.

Let's instead emphasize, again, that a commit and its message should be focused around a particular change and if they have a series of bulleted modifications, those would be better split into multiple commits.

tchaikov replied:
i love explanatory commit message, and i love short commit messages as long as they are well written and they explain the whole story.

sometimes, i just don't have time to make it short.

tchaikov commented on SubmittingPatches.rst:
```diff
@@ -62,468 +66,207 @@ then you just add a line saying ::
+likely the relevant part of the output will have scrolled off your screen or
+gotten swapped out of your buffer. To capture the output for later analysis,
+various tools, such as `script`_ (part of `util-linux`_ which is likely
+already installed on your machine), are available.
```
i don't think we should include this in our document. yes, it's helpful. but it's orthogonal to Ceph development in the sense that it's not closely connected to Ceph or is not specific to Ceph. if we have this hint in this document, shall we include some links to useful Emacs packages or awesome VIM plugins? probably a link to Visual Studio Code as well?

batrick commented on SubmittingPatches.rst:
```diff
@@ -62,468 +66,207 @@ then you just add a line saying ::
-stuff for a multipart patch series. If it's a single patch, don't
-bother.
+The `Title of commit`_ explains that commit message
+titles should look like this::
```
Okay but how should they title the PR? What if they have multiple commits? Then the PR title is usually empty (IIRC) when github opens the PR creation page. We should suggest using the title from the most relevant commit. It is also okay to come up with a completely new title as long as it conforms to the commit title conventions.

batrick replied:
We should also address what the PR body should include. I generally don't spend too much effort on the PR body as most information should be in the commit messages instead. PR body should mostly be used as a center for information about the PR and notices to the maintainers?

tchaikov replied:
agreed. i think PR body is a place for sending messages to reviewer and whoever are interested to the PR. in some cases, we use it as a notepad to 

- document the todo list.
- the performed tests , the test not yet performed.
- the performance benchmarks 
- links for further reading / related PRs

i don't want to codify it. just like the title of an article, the title of a PR should enable its reader to make a decision right away if he/she should take a closer look with a glimpse .

wjwithagen commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+
+If you do not have sufficient permissions to modify any field of the tracker
+issue, just add a comment describing what changes you would like to make.
+Someone with permissions will make the necessary modifications on your behalf.
```
The other way around has worked for me on some occasions: 
 - First create the PR
 - Then create the tracker.
But only then the problem arrises when you forgot to add one release to backport to. 
And you will still have to ask others to fix this for you.

wjwithagen commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,344 @@
+that case, read on.
+
+
+Creating Backport tracker issues
```
What I'm missing here is whether the original master PR needed the `fixes: ....` line before everything below will work??
Especially of those PR's that went into master without a tracker that is hard to fix?

smithfarm replied:
No, you can create the master tracker after the master PR is merged. I edited the document to make this more explicit.

smithfarm commented:
@batrick @tchaikov @wjwithagen Thank you for your reviews. I tried to address all the comments. Please take another look!

smithfarm commented:
Sorry for spamming, but since `SubmittingPatches.rst` is the gateway to Ceph for new contributors, it would be great if more people could review this.

smithfarm commented:
@tchaikov @wjwithagen I tried to address all your review comments. Do you have any further concerns?

LenzGr commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,386 @@
+* cherry-picks must be done using ``git cherry-pick -x``
+* if a commit could not be cherry-picked from master, the commit message must explain why that was not possible
+* the commit message generated by ``git cherry-pick -x`` must not be modified, except to add a "Conflicts" section below the 
+* the "Conflicts" section must mention all files where changes had to made manually (not just conflicts flagged by git)
```
```suggestion
* the "Conflicts" section must mention all files where changes had to be made manually (not just conflicts flagged by git)
```

LenzGr commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,386 @@
+* the commit message generated by ``git cherry-pick -x`` must not be modified, except to add a "Conflicts" section below the 
+* the "Conflicts" section must mention all files where changes had to made manually (not just conflicts flagged by git)
+* the "Conflicts" section should also describe the manual changes that were made
+* if a change is to be backported to multiple stable branches, a tracker issue is needed
```
"...is needed for each branch" maybe?

smithfarm replied:
This is where folks often get confused. They think they need to manually create an issue in the Backport tracker. What is actually needed is an issue in the Bug tracker referring to the master PR and with a populated Backport field. This is explained in the `Tracker workflow` section.

LenzGr replied:
Got it. Makes sense!

LenzGr commented on SubmittingPatches-backports.rst:
```diff
@@ -0,0 +1,386 @@
+Merging a fix directly to a stable (release) branch potentially creates
+a regression in the more recent branches (i.e. all later releases and
+``master``), because it's quite possible the issue will still be present in the
+more recent codestreams after such a fix gets merged to an older stable branch.
```
Not sure I understand this paragraph. If the fix only applies to an older branch, how could the issue still be present in the newer ones?

smithfarm replied:
I agree, it's not clear enough. How about this:

What does "all fixes should land in master first" mean? What if I just need to
fix the issue in <release>?

As the person fixing the issue, you are required to first check whether the issue exists in master. If it does, then the proper course of action is to create a master tracker (see `Tracker workflow`_) and fix the issue in master, first, and only then cherry-pick the fix to the stable branches that have the issue.

If the issue only exists in the stable branch, and not in master, there are several possibilities:

1. it's a regression introduced into the stable branch by a bad backport
2. the issue was fixed in master by some massive refactoring that cannot be backported
3. the issue was already fixed in master by a cherry-pickable commit

In cases 1 and 2, it's permissible to fix the issue directly in the stable branch, subject to the rule "if a commit could not be cherry-picked from master, the commit message must explain why that was not possible".

In case 3, the issue should be handled like any other backport.

LenzGr replied:
That sounds much better, thanks!

LenzGr commented on SubmittingPatches-kernel.rst:
```diff
@@ -0,0 +1,301 @@
+==============================================
```
I believe RST does not require markup before and after a line to mark it as a headline.

LenzGr commented on SubmittingPatches-kernel.rst:
```diff
@@ -0,0 +1,301 @@
+==============================================
+Submitting Patches to Ceph - kernel components
```
```suggestion
Submitting Patches to Ceph - Kernel Components
```

LenzGr commented on SubmittingPatches.rst:
```diff
@@ -62,468 +66,237 @@ then you just add a line saying ::
+notices to maintainers, links to tracker issues and other related information,
+to-do lists, etc. The PR title and description should give the reader a
+high-level notion of what the PR is about, quickly enabling her to decide
+whether she should take a closer look.
```
I suggest replacing "her" with the more neutral singular "they".

smithfarm commented:
Thanks everyone for reviewing. There is still a lot that can be improved. I'm in the process of making further revisions - especially to `SubmittingPatches-backports.rst`, but I moved them to a new branch and will open a follow-up PR. I reverted this branch to ec6af09384be025d5655dc60f5edb717f180b93f - the most recently reviewed SHA1.
